### PR TITLE
confdgnmi - restructure robot test suites

### PR DESCRIPTION
**Pull request by jomiklos**

### CODE REVIEW

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    ...                receives OK response and contains at least one or more items.
+    [Tags]  list
+    Verify Get of  /interfaces/interface
+    ${count}=  Get last updates count
+    Should Not Be Equal As Integers   ${count}  0
+
+Read existing "/interfaces/interface" list entries one by one
+    [Documentation]    Verify that ``GetRequest`` with ``path=interfaces/interface[name]``
+    ...                parameter receives OK response for all/any of the supported Encoding values.
+    [Tags]  list
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot]: The documentation says "for all encodings", so `encoding` tag too?

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    # non-root namespace
+    interfaces/openconfig-interfaces:interface
+    interfaces/openconfig-interfaces:interface/
+    /interfaces/openconfig-interfaces:interface
+    /interfaces/openconfig-interfaces:interface/
+    # both namespaces
+    openconfig-interfaces:interfaces/openconfig-interfaces:interface
+    /openconfig-interfaces:interfaces/openconfig-interfaces:interface
+    openconfig-interfaces:interfaces/openconfig-interfaces:interface/
+    /openconfig-interfaces:interfaces/openconfig-interfaces:interface/
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot]: This is not a good test case. There is no standardization as to how paths should look like or if and how prefixes should be used. For example, if the device requires prefixes (e.g. because it implements both IETF and OC interfaces - weird, but valid), the test case necessary fails. Similarly, if the device does not use prefixes at all because it does not need to, the test case fails too. Also, there is no difference between `/interfaces`, `/interfaces/`, or `interfaces` - slashes are interpreted by the client.

File: confdgnmi/src/confd_gnmi_common.py [hunk 3]
```diff
@@ -192,20 +189,35 @@ def remove_path_prefix(path, prefix):
...
+    return _convert_enum_format(gnmi_pb2.Encoding.Name, encoding, f'Unknown Encoding! ({encoding})',
+                                no_error, -1)
+
+def encoding_int_to_str(encoding: int, no_error=False) -> str:
+    """ Convert integer representation of Encoding to standardized string. """
+    return _convert_enum_format(gnmi_pb2.Encoding.Value, encoding,
+                                f'Unknown Encoding value! ({encoding})',
+                                no_error, f'UNKNOWN({encoding})')
+
+def subscription_str_to_int(mode: str, no_error=False) -> int:
```
Comment (micnovak) [inline confdgnmi/src/confd_gnmi_common.py]: `subscriptionmode_str_to_int`

File: confdgnmi/src/confd_gnmi_common.py [hunk 4]
```diff
@@ -192,20 +189,35 @@ def remove_path_prefix(path, prefix):
...
+    """ Convert text representation of SubscriptionList to standardized integer. """
+    return _convert_enum_format(gnmi_pb2.SubscriptionList.Name, mode,
```
Comment (micnovak) [inline confdgnmi/src/confd_gnmi_common.py]: `gnmi_pb2.SubscriptionList.Name` exists?  How is ti related to `Mode` ??

File: confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    [Documentation]    Get the list of supported encodings from server using ``CapabilityRequest``.
+    ...                Verify that the mandatory JSON encoding is included as one of ``supported_encodings`` values.
+    [Tags]    encoding
+    Get capabilities from device
+    Should Received Ok Response
+    @{supported_encodings}=    List supported encodings
+    List should contain value  ${supported_encodings}  JSON
+
+Supported encodings include JSON or JSON_IETF
+    [Documentation]  Test if the device supports at least one onf JSON/JSON_IETF encodings.
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot]: `one of`

File: confdgnmi/testtool/gNMI_Interface/Capabilities.resource [hunk 6]
```diff
@@ -0,0 +1,13 @@
+*** Settings ***
+Documentation    Resources specific to gNMI ``Capabilities`` RPC/operation.
+Library          CapabilitiesLibrary.py
+
+
+*** Keywords ***
+
+Supported encodings should have some JSON
+    @{encodings}=  List supported encodings
+    Should contain  ${encodings}  JSON_IETF
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/Capabilities.resource]: Is `JSON_IETF` mandatory in our tests? And why?

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    Should Be Equal    ${INTERFACE}    TODO
+
+Get "CONFIG" response includes "config" container
+    [Documentation]    Verify that ``GetRequest`` with ``type=CONFIG`` parameter receives
+    ...                OK response that does include the read/write YANG "config" container.
+    [Tags]  config  container
+    Set Datatype To    CONFIG
+    Verify Get of  /interfaces/interface[name=${INTERFACE}]
+    Check Last Updates Include    config
+
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:105]: What about different set of keywords, using TDD keywords? Something like
```
Given Datatype set to    CONFIG
When Verifying get of    ...
Then Last update includes    config
```
Sounds more like a sentence and may sound familiar to people with some exposure to TDD/BDD (I guess) and looks at least a little bit more side-effect free.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+Get "config" response includes significant leaves
+    [Documentation]    Verify that ``GetRequest`` with ``type=CONFIG`` parameter receives
+    ...                OK response that does include standardized leaf names/items.
+    [Tags]  config  leaf
+    [Template]    Get interface "config" includes leaf
+    name
+    type
+    mtu
+    loopback-mode
+    description
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: These tests differ little bit too much from the rest - in the other ones you explicitly set datatype, invoke get, and do something with the result, here you do it all in one keyword.
Perhaps more importantly - should we really fail the test if `description` is not set? I'd say only `name` and `type` are required, the other ones... I doubt it.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    enabled
+    ifindex
+    admin-status
+    oper-status
+
+Get "config" works for all supported encodings
+    [Documentation]    Verify that ``GetRequest`` with ``type=CONFIG``
+    ...                and ``path=interfaces/interface[]/config`` parameters receives
+    ...                OK response for all/any of the supported ``encoding` parameter values.
+    [Tags]  config  container  encoding
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:169]: IDK, there already is "Get works for all encodings", does this add something substantial?

File: confdgnmi/testtool/OpenConfig/interfaces.resource [hunk 10]
```diff
@@ -0,0 +1,29 @@
+*** Settings ***
+Documentation    Resources specific to ``openconfig-interfaces.yang`` model.
+Library          ../gNMI_Interface/GetLibrary.py    ${DEFAULT_ENCODING}
+Library    Collections
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/interfaces.resource:4]: I don't see where the library is used.

File: confdgnmi/testtool/OpenConfig/openconfig.resource [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+*** Keywords ***
+
+Verify Get of
+    [Arguments]  ${path}
+    [Documentation]    Send ``GetRequest`` with specified path parameter
+    ...                and verify that "ok" response is received.
+    Add Path Parameter  ${path}
+    Dispatch Get Request
+    GetLibrary.Should Received Ok Response
```
Comment (martin-volf) [inline confdgnmi/testtool/OpenConfig/openconfig.resource]: Is this `GetLibrary.` still needed?

File: confdgnmi/testtool/gNMI_Interface/GetLibrary.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                trace(update)
+                updates.append(UpdatePayload.from_obj(update))
+        return updates
+
+    def _updates_include(self, text: str) -> bool:
+        updates = self.get_last_flattened_updates()
+        if updates is None:
+            return False
+        trace(updates)
+        return any(text in update.value for update in updates)
```
Comment (martin-volf) [inline confdgnmi/testtool/gNMI_Interface/GetLibrary.py:138]: I'm afraid this may not be enough - this says that e.g. `"name"` appears as a string in the textual representation of the latest update.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 1 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot]: correct, will add!

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 2 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot]: legit argument, am not familiar with standard in detail, i assumed that both with/without namespace path must be supported for any path even when namespace is not needed due to conflicts...

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 7 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:105]: sounds good, I will rework the test cases...

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 8 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: I agree that value of this may be questionable depending on testing goals.
I assumed tester wants to check "compliance" with standard YANG, thus included all the leaves.
`[Template]` tests are run as separate "tests" and can succeed/fail independently on each other...
This implies that "failure" for some of the leaves like `description` may be just a warning flag that device deviates from standard YANG. We may need to address our stance on cases like this one. (@waitai ?)

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 9 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:169]: I agree that it smells a bit of "random parameters combination" - i tried to come with possibly typical end-user request/RPC...
Intended "added value" if this test-case was that seemingly core/most important "config true" item declared in standard YANG  is included for "type=CONFIG" request independent of encoding parameter.

File: confdgnmi/testtool/OpenConfig/openconfig.resource (hunk 11 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/openconfig.resource]: it isn't, will fix/remove...

File: confdgnmi/testtool/gNMI_Interface/GetLibrary.py (hunk 12 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/GetLibrary.py:138]: Right, this needs to be made more strict/specific to returned data format.
I should have added TODO/notice for this "draft" implementation.
Will fix before merge...

File: confdgnmi/testtool/OpenConfig/interfaces.resource (hunk 10 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/interfaces.resource:4]: e.g. `Set datatype to`...

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 2 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot]: Paths can  be part of device config (file). And this `Get` test can be based just on generic gNMI `Get`  (which probably is)

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 1 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot]: And check if returned value is according to requested encoding?

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 7 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:105]: Ok, but It depends how people see it, for me it is sometimes more difficult to read.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+Get "CONFIG" response includes "config" container
+    [Documentation]    Verify that ``GetRequest`` with ``type=CONFIG`` parameter receives
+    ...                OK response that does include the read/write YANG "config" container.
+    [Tags]  config  container
+    Set Datatype To    CONFIG
+    Verify Get of  /interfaces/interface[name=${INTERFACE}]
+    Check Last Updates Include    config
+
+Get "CONFIG" response does not include "state" container
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:106]: All 4 tests are basically doing the same thing.  Not sure, but some reuse may be considered. Even though they are short.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 8 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: Do we know what are significant leaves? It may be different for different customers/scenarios/devices. If yes, the is should be part of device config.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 9 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:169]: Robot is not strong in parametrization :-1:

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 14]
```diff
@@ -0,0 +1,195 @@
+*** Settings ***
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:1]: I think all tests in OpenConfig should have OpenCOnfig related tag (e.g. `oc`), so we can run `oc` tests only.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+    [Documentation]    Verify that ``GetRequest`` with ``type=CONFIG``
+    ...                and ``path=interfaces/interface[]/config`` parameters receives
+    ...                OK response for all/any of the supported ``encoding` parameter values.
+    [Tags]  config  container  encoding
+
+Get "state" works for all supported encodings
+    [Documentation]    Verify that ``GetRequest`` with ``type=OPERATIONAL``
+    ...                and ``path=interfaces/interface[]/state`` parameters receives
+    ...                OK response for all/any of the supported ``encoding` parameter values.
+    [Tags]  operational  container  encoding
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/01__interfaces.robot:176]: Reuse with previous.

File: confdgnmi/testtool/OpenConfig/interfaces.resource [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+Resource  openconfig.resource
+
+*** Keywords ***
+
+Get container includes
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/interfaces.resource:9]: In which way is this related to OC? This looks like genereral Get that can be re-used also in genral gNMI

File: confdgnmi/testtool/OpenConfig/interfaces.resource [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+Get interface "config" includes leaf
+    # [Arguments]    ${interface}  ${leaf_name}
+    # Get container includes  /interfaces/interface['${interface}']/config  ${leaf_name}
+    # TODO - interface
+    [Arguments]    ${leaf_name}
+    Set Datatype To    CONFIG
+    Get container includes  /interfaces/interface[name=${INTERFACE}]/config  ${leaf_name}
+
+Get interface "state" includes leaf
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/interfaces.resource:20]: Can there be a reuse in `resource` file.

File: confdgnmi/testtool/OpenConfig/openconfig.resource [hunk 18]
```diff
@@ -0,0 +1,35 @@
+*** Settings ***
+Documentation    Resources specific to OpenConfig telemetry testing.
+Library          ../gNMI_Interface/GetLibrary.py    ${DEFAULT_ENCODING}
+
+*** Keywords ***
+
+Verify Get of
```
Comment (micnovak) [inline confdgnmi/testtool/OpenConfig/openconfig.resource:7]: All resources are not related to Openconfig

File: confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    ...              This is a relaxed test in case that device does not follow the "if JSON_IETF is supported,
+    ...              this implies JSON as it's superset automatically" theory.
+    Get capabilities from device
+    Supported encodings should have some JSON
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot]: `some JSON` looks strange and may cause confuision. Maybe  `should have JSON or JSON_IETF`

File: confdgnmi/testtool/gNMI_Interface/02__Get.robot [hunk 20]
```diff
@@ -1,37 +1,37 @@
...
     ...                ignoring the actual payload.
     [Tags]    sanity
     Dispatch Get Request
-    Should receive OK Response
+    Should received OK Response
+
 
 Parameter "prefix" on root path
+    [Documentation]    Try getting whole config by setting ``prefix=/`` parameter.
+    ...                Check that OK response is received without any internal data verification.
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/02__Get.robot:25]: As you say, this may be costly. Perhaps this is good for robustness test, but not for quick `Get` test.

File: confdgnmi/testtool/gNMI_Interface/02__Get.robot [hunk 21]
```diff
@@ -42,55 +42,55 @@ Parameter "DataType" - valid values return OK response
...
-    Verify Get with DataType  INVALID  Should receive Error Response
-
-Parameter "Encoding" - mandatory JSON is advertised
-    [Documentation]    Get the list of supported encodings from server using ``CapabilityRequest``.
-    ...                Verify that the mandatory JSON encoding is included.
-    [Tags]    encoding
-    @{supported}=    Get Supported Encodings
-    List should contain value  ${supported}  JSON
+Parameter "type" - invalid value returns Error response
+    Verify Get with DataType  INVALID  GetLibrary.Should received Error Response
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/02__Get.robot]: This may give impression, that there exists `DataType.INVALID`

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+Suite Setup      Setup gNMI Client
+Suite Teardown   Close gNMI Client
+
+
+*** Test Cases ***
+
+Basic subscription with "mode" parameter
+    [Tags]    sanity
+    [Documentation]  Test that the device correctly responds to all three
+    ...    "Subscribe" request modes.  No further functionality (such as actually
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:17]: No really Subscribe mode, but mode of `SubscriptionList` I think sanity check should be simple.
In case o `STREAM`, aren't we worried it may get not correctly implemented  device into 
bad state and influence other tests. I have feeling this kind of test might be robustness test.

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+#   Path path = 1;                    // The data tree path.
+#   SubscriptionMode mode = 2;        // Subscription mode to be used.
+#   uint64 sample_interval = 3;       // ns between samples in SAMPLE mode.
+#   bool suppress_redundant = 4;
+#   uint64 heartbeat_interval = 5;
+# }
+
+Subscribe ONCE with supported "encoding" values
+
+Two Subscriptions in single request - with same "path"
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:38]: Robustness, concurrency test, not basic gNMI

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+
+Two Subscriptions in single request - with different "path"
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:40]: Strange wording.

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+    # Test is passed path to the data model, which contains limited number of elements. ONCE Subscription operation is invoked with updates_only field set in SubscriptionRequest. Test verifies only one SubscribeResponse with only sync_response set to true.
+
+Basic functionality of the Subscribe POLL RPC .
+    # Parameters: common connection parameters, path, poll count, poll interval.
+    # Failure: Device does not respond, responds with an error, responds with an empty notification set or with a notification without updates, responds with incorrect encoding.
+    # Test is passed path to the data model, which contains limited number of elements. Subscription operation is invoked. First SubscriptionRequest is ONCE. After that POLL subscription requests are invoked with poll interval delay. Test verifies for each subscription data is received. Optionally, it can verify the data is the same as the one received in response for ONCE subscription. Test issues WARNING, if SubscriptionResponse stream is closed prematurely. Test issues WARNING if Updates in SubscriptionResponse are aggregated.
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:79]: Not fully correct wording. Should be ` First SubscriptionRequest is similar to ONCE, with POLL mode in the SubscriptionList, After that POLL SubscriptionRequests messages are ...` - my error in initial text.

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+
+Subscribe POLL RPC with updates_only in the SubscriptionList.
+    # Parameters: common connection parameters, path, poll count, poll interval.
+    # Failure: Device does not respond, responds with an error, responds with an empty notification set or with a notification without updates, responds with incorrect encoding.
+    # Test is passed path to the data model, which contains limited number of elements. Subscription` operation is invoked. First SubscriptionRequest is ONCE with filled SubscriptionList containing updates_only set to true. This subscription is handled as ONCE subscription. After that, empty POLL subscription requests are invoked with poll interval delay. Test verifies for each subscription (also for initial ONCE) a SubscriptionResponse is received with only sync_response set to true (without other fields). Test issues WARNING, if SubscriptionResponse stream is closed prematurely. Test issues WARNING if Updates in SubscriptionResponse are aggregated.
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:84]: Same incorrect wording as previous

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+
+
+Basic functionality of the Subscribe STREAM RPC with ON_CHANGE mode.
+    # Parameters: common connection parameters, path, read count
+    # Failure: Device does not respond, responds with an error, responds with an empty notification set or with a notification without updates, responds with incorrect encoding.
+    # Test is passed path to the data model, which contains limited number of elements, where some of them periodically change (e.g. packet count on interface). STREAM` Subscription operation is invoked. First SubscriptionResponse contains all elements, next responses contains only changed elements. After read count parameter test ends (and subscription stream is closed). Test issues WARNING, if SubscriptionResponse stream is closed prematurely. Test issues WARNING if Updates in SubscriptionResponse are aggregated.
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:90]: Maybe first implementation should end with error instead of WARNING as this is error.

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+    # STREAM SAMPLE mode, with suppress_redundant
+
+Subscription test for all device supported encodings, for unsupported encoding, for non existing encoding.
+
+Subscribe for not existing path
+
+Subscribe for not existing prefix
+
+We will not test TARGET_DEFINED mode
+# TODO - why?
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:112]: Too much device specific

File: confdgnmi/testtool/gNMI_Interface/GetLibrary.py [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+
+@dataclass
+class UpdatePayload:
+    path: str
+    value: Dict[str, object]
+
+    @staticmethod
+    def from_obj(updateObj):
+        path = _make_string_path(updateObj.path, xpath=True)
+        value = json.loads(updateObj.val.json_ietf_val)
```
Comment (micnovak) [inline confdgnmi/testtool/gNMI_Interface/GetLibrary.py:45]: Is this supposed to work only for JSON_IETF encoding?

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 2 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot]: I agree that slashes are redundant due to client parsing, and I will remove them.
Other path cases are still useful IMHO - and can be useful indication of device's gNMI capabilities/implementation details. - It can show whether device can/not use namespaces in paths etc.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 8 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: If user does not need some leaves he can skip specific tests and/or ignore their failure results. 
If user wants 100% OpenConfig compliance, presence of all the leaves is to be expected.

Failing test case does not mean that device is "bad", it just shows what was tested, and what succeeded/failed.
Test case helps to show what target gNMI server device is capable of, and listing negative but reasonably expectable use-cases is IMHO useful as well.

Making any extra things configurable raises knowledge requirement on the test runner/user. This can be problematic due to forcing user to dig into potentially non-interesting details for every other test-case.

Assertion can be changed to warning of logging of all (non)supported leaves as necessarry, we just need to state general approach to compliance.

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 9 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:169]: To answer the initial question -> generic test reaches for root of data tree, while this one specifically for the "CONFIG" node, thus potentially logical difference that could result in other response (assuming server is not ideally implemented).

This test-case definitely falls into "potentially unnecessary" group, but can as well spot one type of bugs in server implementation IMHO.

File: confdgnmi/src/confd_gnmi_common.py (hunk 3 above)
Comment (jomiklos) [reply confdgnmi/src/confd_gnmi_common.py]: :+1: renamed

File: confdgnmi/src/confd_gnmi_common.py (hunk 4 above)
Comment (jomiklos) [reply confdgnmi/src/confd_gnmi_common.py]: this was indeed a bug - fixed in latest codebase

File: confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot (hunk 5 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot]: fixed

File: confdgnmi/testtool/gNMI_Interface/Capabilities.resource (hunk 6 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/Capabilities.resource]: good question 😉  @martin-volf - as initial author of this test case can chime in?

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 13 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:106]: on "backend" - it does reuse code pieces, on end-user test level, it would lead to very short parametrized phrases that are harder to ingest by non-developers IMHO...

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 14 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:1]: There is a tag already for all of them - `OpenConfig`.
I will ad the tag overload also to Openconfig's directory suite setup file (just to be sure).

File: confdgnmi/testtool/OpenConfig/interfaces.resource (hunk 16 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/interfaces.resource:9]: It is generic indeed. I will need to split common methods into standalone resource file without library inclusion.
(to allow using it in places that have different `gNMIRobotLibrary` instances)

File: confdgnmi/testtool/gNMI_Interface/Capabilities.resource (hunk 6 above)
Comment (martin-volf) [reply confdgnmi/testtool/gNMI_Interface/Capabilities.resource]: Because that is the only encoding that really makes sense.  All other ones are underspecified and device specific.

Comment (micnovak) [reply confdgnmi/testtool/gNMI_Interface/Capabilities.resource]: can we be sure gNMI devices behave this way?
https://doc.gambitcom.com/mimic/gnmi.htm

e.g.  this simulator claims JSON is mandatory (only?)

File: confdgnmi/testtool/OpenConfig/openconfig.resource (hunk 18 above)
Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/openconfig.resource:7]: I agree, this is same case as in `interfaces.resource` - potential problem with inheritance/libraries and instance mixups...
will have to rethink how we stucture files / functionality...

File: confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot (hunk 19 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/01__Capabilities.robot]: I agree, updated in latest codebase

File: confdgnmi/testtool/gNMI_Interface/02__Get.robot (hunk 21 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/02__Get.robot]: good point, will clarify within documentation/update name...

File: confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot (hunk 22 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/04__Subscribe.robot:17]: @martin-volf  - ping for your work on subscribe tests...

File: confdgnmi/testtool/OpenConfig/01__interfaces.robot (hunk 8 above)
Comment (micnovak) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: There may be default config. But the main question is how you know the chosen leafs are significant?

Comment (jomiklos) [reply confdgnmi/testtool/OpenConfig/01__interfaces.robot:148]: that's up to each test user to either skip or ignore leaves from standard that are not "significant" for their use-case :)

File: confdgnmi/testtool/gNMI_Interface/GetLibrary.py (hunk 29 above)
Comment (jomiklos) [reply confdgnmi/testtool/gNMI_Interface/GetLibrary.py:45]: this is bug, will fix...
